optaplanner-core/src/main/java/org/optaplanner/core/impl/score/buildin/bendable/BendableScoreInliner.java

```diff
@@ -55,31 +53,27 @@ public IntWeightedScoreImpacter buildWeightedScoreImpacter(BendableScore constra
             int levelWeight = constraintWeight.getHardOrSoftScore(singleLevel);
             if (singleLevel < constraintWeight.getHardLevelsSize()) {
                 int level = singleLevel;
-                return (int matchWeight, Consumer<Score<?>> matchScoreConsumer) -> {
+                return (int matchWeight, Supplier<List<Object>> justifications) -> {
                     int hardImpact = levelWeight * matchWeight;
                     this.hardScores[level] += hardImpact;
-                    if (constraintMatchEnabled) {
-                        matchScoreConsumer
-                                .accept(BendableScore.ofHard(hardScores.length, softScores.length, level, hardImpact));
-                    }
-                    return () -> this.hardScores[level] -= hardImpact;
+                    return buildUndo(constraintPackage, constraintName, constraintWeight,
+                            () -> this.hardScores[level] -= hardImpact,
+                            () -> BendableScore.ofHard(hardScores.length, softScores.length, level, hardImpact),
+                            justifications);
```

Comment:
This is a pattern that will repeat in all `ScoreInliner`s.
All of the code duplication is moved to a new method `buildUndo()`, which deals with constraint matches when they're necessary.